### ISSUE TITLE
Constructing Atomic Subviews

### ISSUE THREAD

**Issue by flying-tiger** [kokkos/kokkos#3765]
I may have run into a bug with the `subview` overload that allows adding memory traits to the source View. Specifically, the subview function template fails to compile when the memory trait is `Atomic`. All other memory traits work. Minimal reproducer:

```
#include <Kokkos_Core.hpp>
int main() {
    using namespace Kokkos;
    View<double*> a("a",10);
    auto range = std::pair(0,4);
    auto s0 = subview(a, range);
    auto s1 = subview<MemoryTraits<Unmanaged>>(a, range);
    auto s2 = subview<MemoryTraits<RandomAccess>>(a, range);
    auto s3 = subview<MemoryTraits<Atomic>>(a, range);   // ERROR!
    auto s4 = subview<MemoryTraits<Restrict>>(a, range);
    auto s5 = subview<MemoryTraits<Aligned>>(a, range);
}
```

Compiler invocation and output:

```
cd /home1/jphill4/projects/nflow/optimize_source/build_v100_nvcc/miniapps/source_miniapp && /swbuild/tsa/apps/gpu_tools/1.2_nvcc/bin/nvcc_wrapper   -Xcudafe --diag_suppr
ess=esa_on_defaulted_function_ignored -O3 -DNDEBUG -fPIE -march=skylake-avx512 -mtune=skylake-avx512 -mrtm -expt-extended-lambda -Wext-lambda-captures-this -expt-relaxed
-constexpr -arch=sm_70 -Xcompiler -fopenmp -std=c++17 -o CMakeFiles/subview.dir/subview.cpp.o -c /home1/jphill4/projects/nflow/optimize_source/miniapps/source_miniapp/su
bview.cpp
/nobackupp16/swbuild/tsa/apps/gpu_tools/1.2_nvcc/include/impl/Kokkos_ViewMapping.hpp(2681): error: expression must have class type
          detected during:
            instantiation of "Kokkos::Impl::ViewDataHandle<Traits, std::enable_if<<expression>, void>::type>::handle_type Kokkos::Impl::ViewDataHandle<Traits, std::enabl
e_if<<expression>, void>::type>::assign(const SrcHandleType &, size_t) [with Traits=Kokkos::ViewTraits<double *, Kokkos::Cuda::array_layout, Kokkos::Device<Kokkos::CudaS
pace::execution_space, Kokkos::CudaSpace::memory_space>, Kokkos::MemoryTraits<6U>>, SrcHandleType=double *]"
(3747): here
            instantiation of "void Kokkos::Impl::ViewMapping<std::enable_if<<expression>, void>::type, SrcTraits, Args...>::assign(Kokkos::Impl::ViewMapping<DstTraits, v
oid> &, const Kokkos::Impl::ViewMapping<SrcTraits, void> &, Args...) [with SrcTraits=Kokkos::ViewTraits<double *>, Args=<std::pair<int, int>>, DstTraits=Kokkos::ViewTrai
ts<double *, Kokkos::Cuda::array_layout, Kokkos::Device<Kokkos::CudaSpace::execution_space, Kokkos::CudaSpace::memory_space>, Kokkos::MemoryTraits<6U>>]"
/nobackupp16/swbuild/tsa/apps/gpu_tools/1.2_nvcc/include/Kokkos_View.hpp(1493): here
            instantiation of "Kokkos::View<DataType, Properties...>::View(const Kokkos::View<RT, RP...> &, Arg0, Args...) [with DataType=double *, Properties=<Kokkos::Cu
da::array_layout, Kokkos::Device<Kokkos::CudaSpace::execution_space, Kokkos::CudaSpace::memory_space>, Kokkos::MemoryTraits<6U>>, RT=double *, RP=<>, Arg0=std::pair<int,
 int>, Args=<>]"
/nobackupp16/swbuild/tsa/apps/gpu_tools/1.2_nvcc/include/Kokkos_View.hpp(1904): here
            instantiation of "Kokkos::Impl::ViewMapping<void, Kokkos::ViewTraits<D, P...>, Args...>::apply<MemoryTraits>::type Kokkos::subview<MemoryTraits,D,P...,Args..
.>(const Kokkos::View<D, P...> &, Args...) [with MemoryTraits=Kokkos::MemoryTraits<6U>, D=double *, P=<>, Args=<std::pair<int, int>>]"
/home1/jphill4/projects/nflow/optimize_source/miniapps/source_miniapp/subview.cpp(12): here

1 error detected in the compilation of "/home1/jphill4/projects/nflow/optimize_source/miniapps/source_miniapp/subview.cpp".
miniapps/source_miniapp/CMakeFiles/subview.dir/build.make:81: recipe for target 'miniapps/source_miniapp/CMakeFiles/subview.dir/subview.cpp.o' failed
```

**Comment by ajpowelsnl** [comment]
@nliber - what is the status of this issue?

**Comment by dalg24** [comment]
Fixed in Kokkos 3.4
